File: apps/demo-be-fsharp-giraffe/tests/DemoBeFsgi.Tests/Integration/Steps/HealthSteps.fs
module DemoBeFsgi.Tests.Integration.Steps.HealthSteps

open System.Text.Json
open TickSpec
open Xunit
open DemoBeFsgi.Tests.State
open DemoBeFsgi.Tests.DirectServices

[<When>]
let ``an operations engineer sends GET /health`` (state: StepState) =
    let status, body = health ()

    { state with
        Response = Some { Status = status; Body = body }
        ResponseBody = Some body }

[<When>]
let ``an unauthenticated engineer sends GET /health`` (state: StepState) =
    let status, body = health ()

    { state with
        Response = Some { Status = status; Body = body }
        ResponseBody = Some body }

[<Then>]
let ``the health status should be "(.+)"`` (status: string) (state: StepState) =
    let doc = JsonDocument.Parse(state.ResponseBody.Value)
    let actual = doc.RootElement.GetProperty("status").GetString()
    Assert.Equal(status, actual)
    state

[<Then>]
let ``the response should not include detailed component health information`` (state: StepState) =
    let doc = JsonDocument.Parse(state.ResponseBody.Value)
    let mutable hasComponents = false

    for prop in doc.RootElement.EnumerateObject() do
        if prop.Name = "components" || prop.Name = "details" then
            hasComponents <- true

    Assert.False(hasComponents, "Response should not include component details")
    state

File: apps/demo-be-python-fastapi/tests/integration/steps/health_steps.py
"""BDD step definitions for health check feature."""

from fastapi.testclient import TestClient
from pytest_bdd import scenarios, then, when

from tests.integration.conftest import GHERKIN_ROOT

scenarios(str(GHERKIN_ROOT / "health" / "health-check.feature"))


@when("an operations engineer sends GET /health", target_fixture="response")
def get_health(client: TestClient):  # type: ignore[no-untyped-def]
    return client.get("/health")


@when("an unauthenticated engineer sends GET /health", target_fixture="response")
def get_health_unauthenticated(client: TestClient):  # type: ignore[no-untyped-def]
    return client.get("/health")


@then('the health status should be "UP"')
def check_health_status(response) -> None:  # type: ignore[no-untyped-def]
    body = response.json()
    assert body.get("status") == "UP", f"Expected status=UP, got: {body}"


@then("the response should not include detailed component health information")
def check_no_component_details(response) -> None:  # type: ignore[no-untyped-def]
    body = response.json()
    assert "components" not in body
    assert "details" not in body
    assert set(body.keys()) == {"status"}, f"Unexpected keys: {set(body.keys())}"
